src/transformers/models/roformer/modeling_roformer.py

```diff
@@ -0,0 +1,1505 @@
+            Whether or not to return the attentions tensors of all attention layers. See ``attentions`` under returned
+            tensors for more detail.
+        output_hidden_states (:obj:`bool`, `optional`):
+            Whether or not to return the hidden states of all layers. See ``hidden_states`` under returned tensors for
+            more detail.
+        return_dict (:obj:`bool`, `optional`):
+            Whether or not to return a :class:`~transformers.file_utils.ModelOutput` instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    "The bare RoFormer Model transformer outputting raw hidden-states without any specific head on top.",
+    ROFORMER_START_DOCSTRING,
+)
+class RoFormerModel(RoFormerPreTrainedModel):
+    """
+
+    The model can behave as an encoder (with only self-attention) as well as a decoder, in which case a layer of
+    cross-attention is added between the self-attention layers, following the architecture described in `Attention is
+    all you need <https://arxiv.org/abs/1706.03762>`__ by Ashish Vaswani, Noam Shazeer, Niki Parmar, Jakob Uszkoreit,
+    Llion Jones, Aidan N. Gomez, Lukasz Kaiser and Illia Polosukhin.
+
+    To behave as an decoder the model needs to be initialized with the :obj:`is_decoder` argument of the configuration
+    set to :obj:`True`. To be used in a Seq2Seq model, the model needs to initialized with both :obj:`is_decoder`
+    argument and :obj:`add_cross_attention` set to :obj:`True`; an :obj:`encoder_hidden_states` is then expected as an
+    input to the forward pass.
+    """
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.config = config
+
+        self.embeddings = RoFormerEmbeddings(config)
+        self.encoder = RoFormerEncoder(config)
+
+        self.init_weights()
+
+    def get_input_embeddings(self):
+        return self.embeddings.word_embeddings
+
+    def set_input_embeddings(self, value):
+        self.embeddings.word_embeddings = value
+
+    def _prune_heads(self, heads_to_prune):
+        """
+        Prunes heads of the model. heads_to_prune: dict of {layer_num: list of heads to prune in this layer} See base
+        class PreTrainedModel
+        """
+        for layer, heads in heads_to_prune.items():
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
+    def get_sinusoidal_position_embeddings(self, inputs):
+        output_dim = self.config.hidden_size // self.config.num_attention_heads
+        position_ids = torch.arange(0, inputs.size(1), dtype=inputs.dtype, device=inputs.device)
+        indices = torch.arange(0, output_dim // 2, dtype=inputs.dtype, device=inputs.device)
+        indices = torch.pow(10000.0, -2 * indices / output_dim)
+        embeddings = torch.einsum('n,d->nd', position_ids, indices)
+        embeddings = torch.stack([embeddings.sin(), embeddings.cos()], dim=-1)
+        embeddings = embeddings.flatten(1, 2)
+        return embeddings[None, None, :, :]
```

Comment:
Could we actually refactor this into a module so the `embedding` tensor can be saved instead of re-computing it for every batch as is done in Marian and m2m_100

https://github.com/huggingface/transformers/blob/4ce6bcc31095ddb8d4cdd79831217f200c53e801/src/transformers/models/m2m_100/modeling_m2m_100.py#L116

https://github.com/huggingface/transformers/blob/4ce6bcc31095ddb8d4cdd79831217f200c53e801/src/transformers/models/marian/modeling_marian.py#L108